docs: update security guide to mention new CSP functionality (#49561)
angular/angular: aio/content/guide/security.md

To enable CSP, configure your web server to return an appropriate `Content-Security-Policy` HTTP header.
Read more about content security policy at the [Web Fundamentals guide](https://developers.google.com/web/fundamentals/security/csp) on the Google Developers website.

The minimal policy required for a brand-new Angular application is:

<code-example format="none" language="none">

> default-src 'self'; style-src 'self' 'nonce-randomNonceGoesHere';

Reviewer: Note: `unsafe-inline` is still a hard requirement until the CSP issue is solved either on Critters side or by extending critters ourselves unless users opt out from critical css inlining. 

From my understanding the person that is assigned on Critters recently returned to work as he was on special leave. So it might take a couple of months to get this resolved.

Author: Isn't it a problem only for the case where link tags are generated? The inline style tags would get the nonce automatically, if it is specified in the index.html.

Reviewer: link tag are always generated.

To enable CSP, configure your web server to return an appropriate `Content-Security-Policy` HTTP header.
Read more about content security policy at the [Web Fundamentals guide](https://developers.google.com/web/fundamentals/security/csp) on the Google Developers website.

The minimal policy required for a brand-new Angular application is:

<code-example format="none" language="none">

default-src 'self'; style-src 'self' 'nonce-randomNonceGoesHere'; script-src 'self' 'nonce-randomNonceGoesHere';

</code-example>

> You have to tell Angular what the **randomly-generated** nonce from the header is so that it can be used when inserting component styles into the page.
> There are two ways to provide the nonce:

Reviewer 2: ```suggestion
When serving your Angular application, the server should include a 
randomly-generated nonce in the HTTP header for each request. 
You must provide this nonce to Angular so that the framework
can render `<style>` elements.

You can set the nonce for Angular in one of two ways:
```

> 1. Set the `ngCspNonce` attribute on the root application element as `<app ngCspNonce="randomNonceGoesHere"></app>`. This approach is useful if you have access to server-side templating that can add the nonce both to the header and the `index.html` when constructing the response.
> 2. Provide the nonce using the `CSP_NONCE` injection token. This can be useful if you have access to the nonce at runtime and you want to be able to cache the `index.html`.

Reviewer 2: ```suggestion
1. Set the `ngCspNonce` attribute on the root application element as `<app ngCspNonce="randomNonceGoesHere"></app>`. Use this approach if you have access to server-side templating that can add the nonce both to the header and the `index.html` when constructing the response.
2. Provide the nonce using the `CSP_NONCE` injection token. Use this approach if you have access to the nonce at runtime and you want to be able to cache the `index.html`.
```

<code-example format="typescript" language="typescript">

import {bootstrapApplication, CSP_NONCE} from '&commat;angular/core';
import {AppComponent} from './app/app.component';

bootstrapApplication(AppComponent, {
  providers: [{
    provide: CSP_NONCE,
    useValue: globalThis.myRandomNonceValue
  }]
});

</code-example>

<div class="callout is-helpful">

> It's important that the nonce you provide to Angular is <strong>unique</strong> and that it can't be guessed.
> If an attacker is able to predict what the nonce will be, they will be able to get around the protections offered by CSP.

Reviewer 2: ```suggestion
Always ensure that the nonces you provide are <strong>unique per request</strong>
and that they are not predictable or guessable. If an attacker can predict future nonces, they can circumvent the protections offered by CSP.
```

</div>

> If providing the nonce is problematic in your project, you can allow inline styles by adding `'unsafe-inline'` to the `style-src` section of the CSP header instead, but note that this can compromise the application's security.
> 

Reviewer 2: ```suggestion
If you cannot generate nones in your project, you can allow inline styles by adding `'unsafe-inline'` to the `style-src` section of the CSP header. 
```

I would leave off the bit about "can compromise the application's security" since we don't actually believe that unsafe-inline style-src actually opens the door to any attacks.

| Sections                | Details |
|:---                     |:---     |
| `default-src 'self';`   | Allows the page to load all its required resources from the same origin. |

> | `style-src 'self' 'nonce-randomNonceGoesHere';`     | Allows the page to load global styles from the same origin \(`'self'`\) and ones inserted by Angular with the `nonce-randomNonceGoesHere`. |

Reviewer 2: ```suggestion
| `style-src 'self' 'nonce-randomNonceGoesHere';`     | Allows the page to load global styles from the same origin \(`'self'`\) and styles inserted by Angular with the `nonce-randomNonceGoesHere`. |
```

> | `script-src 'self' 'nonce-randomNonceGoesHere';`     | Allows the page to load JavaScript from the same origin \(`'self'`\) and ones inserted by the Angular CLI with the `nonce-randomNonceGoesHere`. This is only required if you're using critical CSS inlining. |

Reviewer 2: I don't think we should include `script-src` with a nonce here?

Author: The `script-src` is necessary when using critical CSS inlining. It works by setting `media="print"` on link tags and setting it back to `media="all"` after it has loaded. See https://github.com/angular/angular-cli/pull/24903.
